node/core/pvf/src/host.rs

```diff
@@ -0,0 +1,1039 @@
+//! jobs for PVF execution or preparation.
+//!
+//! The validation host is represented by a future/task that runs an event-loop and by a handle,
+//! [`ValidationHost`], that allows communication with that event-loop.
+
+use crate::{
+	Priority, Pvf, ValidationError,
+	artifacts::{Artifacts, ArtifactState, ArtifactId},
+	execute, prepare,
+};
+use std::{
+	collections::HashMap,
+	time::{Duration, SystemTime},
+};
+use always_assert::never;
+use async_std::{
+	path::{Path, PathBuf},
+	sync::Mutex,
+};
+use polkadot_parachain::primitives::ValidationResult;
+use futures::{
+	Future, FutureExt, SinkExt, StreamExt,
+	channel::{mpsc, oneshot},
+};
+
+/// An alias to not spell the type for the oneshot sender for the PVF execution result.
+pub(crate) type ResultSender = oneshot::Sender<Result<ValidationResult, ValidationError>>;
+
+/// A handle to the async process serving the validation host requests.
+pub struct ValidationHost {
+	to_host_tx: Mutex<mpsc::Sender<ToHost>>,
+}
+
+impl ValidationHost {
+	/// Execute PVF with the given code, params and priority. The result of execution will be sent
+	/// to the provided result sender.
+	///
+	/// This is async to accomodate the fact a possibility of back-pressure. In vast majority of
+	/// situations this function should return immediatelly.
+	///
+	/// Returns an error if the request cannot be sent to the validation host, i.e. if it shut down.
+	pub async fn execute_pvf(
+		&self,
+		pvf: Pvf,
+		params: Vec<u8>,
+		priority: Priority,
+		result_tx: ResultSender,
+	) -> Result<(), String> {
+		self.to_host_tx
+			.lock()
+			.await
+			.send(ToHost::ExecutePvf {
+				pvf,
+				params,
+				priority,
+				result_tx,
+			})
+			.await
+			.map_err(|_| "the inner loop hung up".to_string())
+	}
+
+	/// Sends a signal to the validation host requesting to prepare a list of the given PVFs.
+	///
+	/// This is async to accomodate the fact a possibility of back-pressure. In vast majority of
+	/// situations this function should return immediatelly.
+	///
+	/// Returns an error if the request cannot be sent to the validation host, i.e. if it shut down.
+	pub async fn heads_up(&self, active_pvfs: Vec<Pvf>) -> Result<(), String> {
+		self.to_host_tx
+			.lock()
+			.await
+			.send(ToHost::HeadsUp { active_pvfs })
+			.await
+			.map_err(|_| "the inner loop hung up".to_string())
+	}
+}
+
+enum ToHost {
+	ExecutePvf {
+		pvf: Pvf,
+		params: Vec<u8>,
+		priority: Priority,
+		result_tx: ResultSender,
+	},
+	HeadsUp {
+		active_pvfs: Vec<Pvf>,
+	},
+}
+
+/// Configuration for the validation host.
+pub struct Config {
+	/// The root directory where the prepared artifacts can be stored.
+	pub cache_path: PathBuf,
+	/// The path to the program that can be used to spawn the prepare workers.
+	pub prepare_worker_program_path: PathBuf,
+	/// The time alloted for a prepare worker to spawn and report to the host.
+	pub prepare_worker_spawn_timeout: u64,
```

Comment:
It would be easier to read if the var name held the time

Reply:
agreed. Or perhaps we can just plumb `Duration`